### DISCUSSION TITLE
Benchmark Study: Punge (YOLOv11n) On-Device vs. Google Cloud Vision for NSFW Image Detection

### DISCUSSION THREAD

**Discussion by markatlarge** [ultralytics/ultralytics#21540] [Show and tell]
Just published my first benchmark study comparing NSFW detection models, including my own lightweight, on-device model. Would really appreciate any feedback—good, bad, or brutally honest.
Always looking to learn and improve. Thanks in advance to anyone who takes a look. https://medium.com/@russoatlarge/benchmark-study-punge-yolov11n-on-device-vs-google-cloud-vision-for-nsfw-image-detection-d4462217ed15

**Comment by UltralyticsAssistant** [comment]
👋 Hello @markatlarge, thank you for sharing your benchmark study and for your interest in Ultralytics 🚀! We appreciate contributions from the community and welcome feedback, insights, and discussions around YOLO11n and on-device NSFW detection.

If you have any specific 🐛 bug reports or run into issues while implementing or benchmarking YOLO on your device, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us better assist you.

For custom training or model evaluation ❓ questions, including additional details like dataset samples, training logs, or your benchmarking workflow will help the team and community provide more targeted feedback. You can also check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to get the most out of your experiments.

Feel free to engage with the Ultralytics community:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- Or join the conversation on our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

To ensure you’re evaluating with the latest features and fixes, upgrade to the newest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖. An Ultralytics engineer will review your discussion and follow up soon!

**Reply by glenn-jocher** [reply]
Great work on the benchmark study comparing on-device YOLO11n performance with Google Cloud Vision for NSFW detection! It's always valuable to see real-world performance comparisons, especially for edge deployment scenarios.

Your approach of using a lightweight on-device model addresses important considerations around privacy, latency, and cost that many developers face. If you'd like to dive deeper into YOLO benchmarking techniques or share specific performance metrics you encountered, our [benchmarking documentation](https://docs.ultralytics.com/modes/benchmark/) covers various aspects of model evaluation across different hardware setups.

Feel free to share any specific findings or challenges from your study here - the community would likely benefit from your insights on on-device NSFW detection performance!

**Reply by markatlarge** [reply]
Thanks, really appreciate the kind words! I was pretty surprised by the results myself and I’m looking into getting them independently validated.

For NSFW detection, I think YOLO has a lot going for it—especially when running on-device. In this study, I just used a single confidence threshold across all classes, but you could definitely get better performance by tuning it per class. And with bounding boxes, you can go a step further by actually obscuring just the sensitive parts of an image—something you can’t really do with models that only return a label.

Happy to chat more about benchmarking or deployment if it's helpful—always learning, and always open to feedback!

**Reply by glenn-jocher** [reply]
Those are excellent insights, especially about the per-class confidence tuning and leveraging bounding boxes for selective content obscuration - that's a compelling advantage of object detection over simple classification for content moderation use cases.

Your point about independent validation is spot on for benchmark credibility. The ability to fine-tune detection thresholds per class using the `conf` parameter can indeed significantly improve precision for specific NSFW categories where you might want different sensitivity levels.

The bounding box approach for selective obscuration is particularly innovative - being able to blur or mask only the problematic regions while preserving the rest of the image content is a huge practical advantage for many applications. Thanks for sharing these real-world insights with the community!
